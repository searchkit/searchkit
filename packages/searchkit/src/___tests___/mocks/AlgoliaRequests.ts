export const SimpleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequestOneFilter = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['actors:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const DisjunctiveExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'type',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const NestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const TwoFiltersNestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice', 'user.first:John']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFilterNestedQueryRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetFilters: [],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['user.price>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const FacetQueryExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetName: 'type',
      facetQuery: 'm',
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFiltersExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['imdbrating>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      clickAnalytics: false,
      facets: 'imdbrating',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 0,
      maxValuesPerFacet: 10,
      page: 0,
      query: ''
    }
  }
]

export const NestedFacetQueryExampleRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetName: 'user.first',
      facetQuery: 'har',
      facets: ['user.first', 'user.price'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]
