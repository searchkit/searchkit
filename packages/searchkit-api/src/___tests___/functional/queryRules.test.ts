import Client from '../../'
import { DisjunctiveExampleRequest } from '../mocks/AlgoliaRequests'
import type { AlgoliaMultipleQueriesQuery } from '@searchkit/api'

import nock from 'nock'
import { HitsResponseWithFacetFilter } from '../mocks/ElasticsearchResponses'

describe('Integration tests for query rules', () => {
  it('call with one filter and query applied', async () => {
    const client = Client({
      connection: {
        host: 'https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title', 'actors'],
        search_attributes: ['title', 'actors', 'query'],
        result_attributes: ['title', 'actors', 'query'],
        facet_attributes: [
          'type',
          { field: 'actors.keyword', attribute: 'actors', type: 'string' },
          'rated'
        ],
        query_rules: [
          {
            conditions: [
              [
                {
                  context: 'query',
                  match_type: 'exact',
                  value: 'shawshank'
                }
              ]
            ],
            actions: [
              {
                action: 'PinnedResult',
                documentIds: ['1']
              },
              {
                action: 'RenderUserData',
                userData: '{"title":"Movies"}'
              },
              {
                action: 'RenderFacetsOrder',
                facetAttributesOrder: ['type', 'actors']
              }
            ]
          }
        ]
      }
    })

    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        return true
      })
      .reply(200, HitsResponseWithFacetFilter)

    const response = await client.handleInstantSearchRequests(
      DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })
})
