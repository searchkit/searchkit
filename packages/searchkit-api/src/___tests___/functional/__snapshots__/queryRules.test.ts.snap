// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests for query rules call with one filter and query applied 1`] = `
{
  "results": [
    {
      "appliedRules": [
        "1",
      ],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "actors": {
          "Bob Gunton": 1,
          "Morgan Freeman": 1,
          "Tim Robbins": 1,
          "William Sadler": 1,
        },
        "rated": {
          "R": 1,
        },
        "type": {
          "movie": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {
            "actors": [
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Tim Robbins",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Morgan Freeman",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Bob Gunton",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "William Sadler",
              },
            ],
            "title": {
              "fullyHighlighted": false,
              "matchLevel": "full",
              "matchedWords": [
                "Shawshank",
              ],
              "value": "The <em>Shawshank</em> Redemption",
            },
          },
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "type",
              "actors",
            ],
          },
          "values": {
            "actors": {
              "sortRemainingBy": "count",
            },
            "type": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
      "userData": [
        {
          "title": "Movies",
        },
      ],
    },
    {
      "appliedRules": [
        "1",
      ],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "type": {
          "movie": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {
            "actors": [
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Tim Robbins",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Morgan Freeman",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Bob Gunton",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "William Sadler",
              },
            ],
            "title": {
              "fullyHighlighted": false,
              "matchLevel": "full",
              "matchedWords": [
                "Shawshank",
              ],
              "value": "The <em>Shawshank</em> Redemption",
            },
          },
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 1,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "analytics=false&clickAnalytics=false&facets=type&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&hitsPerPage=1&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "type",
              "actors",
            ],
          },
          "values": {
            "actors": {
              "sortRemainingBy": "count",
            },
            "type": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
      "userData": [
        {
          "title": "Movies",
        },
      ],
    },
  ],
}
`;

exports[`Integration tests for query rules call with one filter and query applied: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{"type":{"terms":{"field":"type","size":10}},"actors":{"terms":{"field":"actors.keyword","size":10}}},"query":{"bool":{"filter":[{"bool":{"should":[{"term":{"type":"movie"}}]}}],"must":{"function_score":{"query":{"pinned":{"ids":["1"],"organic":{"bool":{"should":[{"bool":{"should":[{"multi_match":{"query":"shawshank","fields":["title","actors","query"],"fuzziness":"AUTO:4,8"}},{"multi_match":{"query":"shawshank","fields":["title","actors","query"],"type":"bool_prefix"}}]}},{"multi_match":{"query":"shawshank","type":"phrase","fields":["title","actors","query"]}}]}}}},"functions":[]}}}},"size":20,"from":0,"_source":{"includes":["title","actors","query"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0},"actors":{"number_of_fragments":0}}}}
{"index":"imdb_movies"}
{"aggs":{"type":{"terms":{"field":"type","size":10}}},"query":{"bool":{"filter":[],"must":{"function_score":{"query":{"pinned":{"ids":["1"],"organic":{"bool":{"should":[{"bool":{"should":[{"multi_match":{"query":"shawshank","fields":["title","actors","query"],"fuzziness":"AUTO:4,8"}},{"multi_match":{"query":"shawshank","fields":["title","actors","query"],"type":"bool_prefix"}}]}},{"multi_match":{"query":"shawshank","type":"phrase","fields":["title","actors","query"]}}]}}}},"functions":[]}}}},"size":1,"from":0,"_source":{"includes":["title","actors","query"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0},"actors":{"number_of_fragments":0}}}}
"
`;
